app/upgrade_handler.go

```diff
@@ -97,3 +102,68 @@ func (app ShentuApp) setTmpUpgradeHandler() {
 		app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, &storeUpgrades))
 	}
 }
+
+func RunShieldMigration(app ShentuApp, ctx sdk.Context) {
+	sk := app.ShieldKeeper
+	bondDenom := sk.BondDenom(ctx)
+
+	moduleCoins := app.BankKeeper.GetAllBalances(ctx, app.AccountKeeper.GetModuleAccount(ctx, shieldtypes.ModuleName).GetAddress())
+
+	// remaining service fees
+	remainingServiceFees := sk.GetRemainingServiceFees(ctx)
+
+	// rewards
+	var rewards sdk.DecCoins
+	for _, provider := range sk.GetAllProviders(ctx) {
+		rewards = rewards.Add(provider.Rewards...)
+	}
+
+	totalInt, _ := remainingServiceFees.Add(rewards...).TruncateDecimal()
+
+	// shield stake
+	shieldStake := sdk.ZeroInt()
+	for _, stake := range sk.GetAllStakeForShields(ctx) {
+		shieldStake = shieldStake.Add(stake.Amount)
+	}
+
+	// reimbursement
+	reimbursement := sdk.ZeroInt()
+	for _, rmb := range sk.GetAllReimbursements(ctx) {
+		reimbursement = reimbursement.Add(rmb.Amount.AmountOf(bondDenom))
+	}
+
+	// block service fees
+	blockServiceFees, _ := sk.GetBlockServiceFees(ctx).TruncateDecimal()
+
+	totalInt = totalInt.Add(sdk.NewCoin(bondDenom, shieldStake)).Add(sdk.NewCoin(bondDenom, reimbursement)).Add(blockServiceFees...)
+
+	if moduleCoins.IsAllGTE(totalInt) {
+		blockServiceFees = blockServiceFees.Add(moduleCoins.Sub(totalInt)...)
+		newBlockServiceFees := sdk.NewDecCoinsFromCoins(blockServiceFees...)
+		sk.SetBlockServiceFees(ctx, newBlockServiceFees)
+	} else {
+		diff := totalInt.Sub(moduleCoins)
+		fmt.Println("diff: ", diff)
+		// first try to take away from remaining services
+		rSFInt, decimals := remainingServiceFees.TruncateDecimal()
+		if !rSFInt.IsAllGTE(diff) {
+			additionalFunds := diff.Sub(rSFInt)
+			app.BankKeeper.SendCoinsFromModuleToModule(ctx, distrtypes.ModuleName, shieldtypes.ModuleName, additionalFunds)
```

Comment:
Shall we check the coins in distribute module account is more than `additionalFunds` ? 
Please skip this comment if there is an obvious yes answer.

Reply:
I think we can assume that, but I'll add a check.